### PR TITLE
[MRG+1] BF: IndexError in fast_abs_percentile

### PR DESCRIPTION
**Pull request by dohmatob**
Fixes issue #872.

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
In general, LGTM.

However, I would like also a slightly more atomic test of only fast_abs_percentile, as you had in your original issue.

**Comment by lesteve** [comment]
See associated issue, I feel this is a wider issue with fast_abs_percentile than just an out-of-bound indexing issue.

**Comment by GaelVaroquaux** [comment]
LGTM. +1 for merge.

File: nilearn/tests/test_extmath.py [hunk 1]
```diff
@@ -8,11 +8,16 @@
...
 from nilearn._utils.extmath import fast_abs_percentile, is_spd
 
+from sklearn.datasets import load_iris
+from ..image import mean_img
+from ..decoding.tests.test_same_api import to_niimgs
+from ..plotting.find_cuts import find_xyz_cut_coords
+
 
 def test_fast_abs_percentile():
     data = np.arange(1, 100)
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: Can you shuffle it? This is the reason the test were passing before with the completely broken implementation of fast_abs_percentile.

File: nilearn/tests/test_extmath.py [hunk 2]
```diff
@@ -50,3 +55,22 @@ def test_is_spd_with_symmetrical_matrix():
     matrix = np.array([[2, 1],
                        [1, 1]])
     assert is_spd(matrix, verbose=0)
+
+
+def test_fast_abs_percentile_no_index_error():
+    # Regression test
+    #####################
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: Just curious about the human psyche, why do you feel the need to add your own pet markup that is not used anywhere else in the file?

Aside from that, I don't this comments adds a lot.

File: nilearn/tests/test_extmath.py [hunk 3]
```diff
@@ -50,3 +55,22 @@ def test_is_spd_with_symmetrical_matrix():
...
+
+    # check the offending low-level function
+    fast_abs_percentile(np.arange(4))
+
+
+def test_fast_abs_percentile_no_index_error_find_cuts():
+    # Regression test
+    #####################
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: Same thing.

File: nilearn/tests/test_extmath.py [hunk 4]
```diff
@@ -50,3 +55,22 @@ def test_is_spd_with_symmetrical_matrix():
...
+
+    # check that find_cuts functions are safe
+    iris = load_iris()
+    X, _ = iris.data, iris.target
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: You are not using iris at all, right?

File: nilearn/tests/test_extmath.py [hunk 5]
```diff
@@ -50,3 +55,22 @@ def test_is_spd_with_symmetrical_matrix():
...
+    X, _ = to_niimgs(X, [2, 2, 2])
+    nose.tools.assert_equal(len(find_xyz_cut_coords(mean_img(X))), 3)
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: I don't think test_extmath.py is the right place to test find_xyz_cut_coords. The additional imports at the top of the file should be removed too.

File: nilearn/tests/test_extmath.py (hunk 1 above)
Comment (dohmatob) [inline nilearn/tests/test_extmath.py]: On Wed, Dec 2, 2015 at 5:16 PM, Loïc Estève <notifications@github.com>
wrote:

> In nilearn/tests/test_extmath.py
> <https://github.com/nilearn/nilearn/pull/875#discussion_r46435121>:
>
> > @@ -8,11 +8,16 @@
> >
> >  from nilearn._utils.extmath import fast_abs_percentile, is_spd
> >
> > +from sklearn.datasets import load_iris
> > +from ..image import mean_img
> > +from ..decoding.tests.test_same_api import to_niimgs
> > +from ..plotting.find_cuts import find_xyz_cut_coords
> > +
> >
> >  def test_fast_abs_percentile():
> >      data = np.arange(1, 100)
>
> Can you shuffle it? This is the reason the test were passing before with
> the completely broken implementation of fast_abs_percentile.
>
Good catch. Worst still the data array is not even used...

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/875/files#r46435121>.
>



-- 
Elvis Dohmatob,
*sent from my  potato...*
http://dohmatob.blogspot.fr/, https://team.inria.fr/parietal/elvis/,
https://github.com/dohmatob

File: nilearn/tests/test_extmath.py [hunk 6]
```diff
@@ -6,13 +6,16 @@
...
 
 
 def test_fast_abs_percentile():
-    data = np.arange(1, 100)
-    for p in range(10, 100, 10):
-        yield nose.tools.assert_equal, fast_abs_percentile(data, p - 1), p
+    rng = check_random_state(1)
+    data = np.arange(10, 100, 10)
+    rng.shuffle(data)
+    for p in data:
```
Comment (lesteve) [inline nilearn/tests/test_extmath.py]: I am not sure it is a great idea to have `data` and the percentiles we are checking to be identical. Basically you want to check the 10%, 20%, ..., 90% percentiles of `data` and `data` can be anything.

Comment (lesteve) [inline nilearn/tests/test_extmath.py]: In other words, I would still have `data` a shuffle of `np.arange(1, 100)` and `percentiles_to_check = range(10, 100, 10)`.

Comment (dohmatob) [inline nilearn/tests/test_extmath.py]: @Loic: this PR is solely meant to fix an Index error. Those tests existed
independently of the index error. These comments have nothing to do with
current PR (lest, we'll end up having to modify tests in NiftiMasker, e.g)

On Wed, Dec 2, 2015 at 6:07 PM, Loïc Estève <notifications@github.com>
wrote:

> In nilearn/tests/test_extmath.py
> <https://github.com/nilearn/nilearn/pull/875#discussion_r46442948>:
>
> >  from nilearn._utils.extmath import fast_abs_percentile, is_spd
> >
> >
> >  def test_fast_abs_percentile():
> > -    data = np.arange(1, 100)
> > -    for p in range(10, 100, 10):
> > -        yield nose.tools.assert_equal, fast_abs_percentile(data, p - 1), p
> > +    rng = check_random_state(1)
> > +    data = np.arange(10, 100, 10)
> > +    rng.shuffle(data)
> > +    for p in data:
>
> I am not sure it is a great idea to have data and the percentiles we are
> checking to be identical. Basically you want to check the 10%, 20%, ...,
> 90% percentiles of data and data can be anything.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nilearn/nilearn/pull/875/files#r46442948>.
>



-- 
Elvis Dohmatob,
*sent from my  potato...*
http://dohmatob.blogspot.fr/, https://team.inria.fr/parietal/elvis/,
https://github.com/dohmatob

Comment (lesteve) [inline nilearn/tests/test_extmath.py]: > Those tests existed independently of the index error.

:confused: what I am trying to get get at is to leave the tests as they were, except for the shuffle which I think was a big flaw since it gave us false confidence that fast_abs_percentile was working when it wasn't.

Comment (lesteve) [inline nilearn/tests/test_extmath.py]: > what I am trying to get at is to leave the tests as they were

Obviously you do need the 
`nose.tools.assert_equal, fast_abs_percentile(data, p), p`
rather than
`nose.tools.assert_equal, fast_abs_percentile(data, p - 1), p`

to get the tests to pass.

File: nilearn/plotting/tests/test_find_cuts.py [hunk 7]
```diff
@@ -4,6 +4,10 @@
 from nilearn.plotting.find_cuts import (find_xyz_cut_coords, find_cut_slices,
                                         _transform_cut_coords)
 from nilearn._utils.testing import assert_raises_regex, assert_warns
+from sklearn.datasets import load_iris
+from ...image import mean_img
+from ...decoding.tests.test_same_api import to_niimgs
+from ...plotting.find_cuts import find_xyz_cut_coords
```
Comment (GaelVaroquaux) [inline nilearn/plotting/tests/test_find_cuts.py]: We usually use absolute imports in tests, so that they can be run as scripts.

File: nilearn/plotting/tests/test_find_cuts.py [hunk 8]
```diff
@@ -4,6 +4,10 @@
 from nilearn.plotting.find_cuts import (find_xyz_cut_coords, find_cut_slices,
                                         _transform_cut_coords)
 from nilearn._utils.testing import assert_raises_regex, assert_warns
+from sklearn.datasets import load_iris
```
Comment (GaelVaroquaux) [inline nilearn/plotting/tests/test_find_cuts.py]: Would you mind moving this line up, before the nilearn imports, to separate nilearn imports from the rest.

File: nilearn/plotting/tests/test_find_cuts.py [hunk 9]
```diff
@@ -133,3 +137,11 @@ def test_find_cuts_empty_mask_no_crash():
...
                               mask=mask)
     np.testing.assert_array_equal(cut_coords, [.5, .5, .5])
+
+
+def test_fast_abs_percentile_no_index_error_find_cuts():
+    # check that find_cuts functions are safe
+    iris = load_iris()
+    X, _ = iris.data, iris.target
+    X, _ = to_niimgs(X, [2, 2, 2])
+    assert_equal(len(find_xyz_cut_coords(mean_img(X))), 3)
```
Comment (lesteve) [inline nilearn/plotting/tests/test_find_cuts.py]: I can reproduce a similar error with:

```python
import numpy as np

import nibabel

from nilearn.plotting.find_cuts import find_xyz_cut_coords

data = np.array([
    [[1., 2.],
     [3., 4.]],
    [[0., 0.],
     [0., 0.]]])

img = nibabel.Nifti1Image(data, np.eye(4))

find_xyz_cut_coords(img)

```

i.e. no need to resort to load_iris, to_niimgs and mean_img.

**Comment by dohmatob** [comment]
Normally, all comments have been addressed.

**Comment by lesteve** [comment]
> Normally, all comments have been addressed.

This looks great, thanks a lot! Let's wait for the CI servers and hopefully merge this when they come back all green.

**Comment by GaelVaroquaux** [comment]
Thanks Elvis; it's super useful to fix such bugs. They are showstoppers
for non-advanced users. For us, advanced users, they bug us right before
deadlines, and thus make us loose a lot of time.

**Comment by lesteve** [comment]
Test are all green, merging, thanks a lot!
